### PR TITLE
Add SelectDtypeColumns component

### LINKED ISSUE #2479: Create a new component which selects using feature types
We currently have the `SelectColumns` component which allows us to select specific columns based on column names. However, it would be useful to also have a component which filters / selects columns by (Woodwork logical) type. This could help us enable specific subgraphs / pipelines that handle certain data types (ex: select categorical cols --> pass to OHE).

### PR DESCRIPTION
**Pull request by eccabay**
Closes #2479 

I'm totally open to changing the name of the class, it was just the first thing I thought of

### CODE REVIEW

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (eccabay) [inline evalml/pipelines/components/transformers/column_selectors.py:150]: This logic is definitely not the cleanest it could be - if anyone knows a more efficient way to pull out this information, please let me know!

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (freddyaboulton) [inline evalml/pipelines/components/transformers/column_selectors.py:141]: What are your thoughts on making this a no-op? 

If we do that, then we get selection by semantic tag for free. I feel like that would be useful for workflows that don't use the "standard" set of types. For example, a user might be interested in selecting all numeric features (including Age and AgeNullable) or all categorical features (CountryCode and Ordinal). It would also make it easy for users to select based on their own criteria if they set the tags correctly. For example, users may be interested in selecting based on whether a feature was created by another component or not.

Another argument for doing that is that `check_input_for_columns` was implemented because `X.ww[cols]` would error if the columns were not in the dataframe. However, `Xww.select` does not error out if the resulting dataframe is empty. I think it's ok if this returns an empty dataframe - maybe we can just raise a warning instead?

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (freddyaboulton) [inline evalml/pipelines/components/transformers/column_selectors.py:164]: I think this got carried over from the `DropColumns` transformer!

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (freddyaboulton) [inline evalml/pipelines/components/transformers/column_selectors.py:127]: Regarding the name, I feel like the right thing to do refactor `SelectColumns` to allow for logical type selection - that way there's only one `SelectColumns` transformer - but I don't think that's in scope for this issue.

If we follow my suggestion to also allow selecting by semantic tags, I think `SelectByTypeorTag` is a good alternate name. If not, I think `SelectByLogicalType` is also a good name! I don't like `Dtype` because  that's more pandas terminology and I feel it obscures the fact it's done via woodwork.

**Review by freddyaboulton — APPROVED** [review]
@eccabay Looks good! Thanks for using the existing `ColumnSelector` abstraction!

One comment I'd like to resolve before merge is that if we make `_check_input_for_columns` a no-op we get semantic tag selection for free! I think that'll make this more flexible.

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (bchen1116) [reply evalml/pipelines/components/transformers/column_selectors.py:141]: Agreed. I think this could definitely be a no-op, since it does allow us to select by semantic tags automatically.

Additionally,, currently it doesn't allow for selection if you use the capital string logical types (ie "NaturalLanguage" vs "natural_language")
image
I would think this should work since I can call `X.ww.select()` with both and get valid results

File: evalml/tests/component_tests/test_column_selector_transformers.py
Comment (bchen1116) [inline evalml/tests/component_tests/test_column_selector_transformers.py:257]: Is this supposed to remain commented?

File: evalml/tests/component_tests/test_column_selector_transformers.py
Comment (bchen1116) [inline evalml/tests/component_tests/test_column_selector_transformers.py:240]: Can you add a test that selects based on capitalized types? ie `"NaturalLanguage"`, `"Double"`

**Review by bchen1116 — CHANGES_REQUESTED** [review]
Looking good! Agreed with Freddy, might be nice to not limit it to selecting by logical types. Left some comments on additional tests that could be helpful, as well as some clean up

File: evalml/tests/component_tests/test_column_selector_transformers.py
Comment (eccabay) [reply evalml/tests/component_tests/test_column_selector_transformers.py:257]: It is not, thanks for the catch!

**Review by ParthivNaresh — APPROVED** [review]
Looks great! After incorporating Freddy's and Bryan's changes I'd say this is solid. Nice job expanding the test to cover Double as well. I just left a minor comment regarding the docstring.

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (freddyaboulton) [reply evalml/pipelines/components/transformers/column_selectors.py:138]: Great point @ParthivNaresh !!

**Review by bchen1116 — APPROVED** [review]
Nice! LGTM!

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (chukarsten) [inline evalml/pipelines/components/transformers/column_selectors.py:153]: Do we want this behavior?  I would think that we might want to take the same route as the sibling classes and raise an exception if a bad selection is made.  @angela97lin what use cases are we trying to satisfy with this?

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (chukarsten) [inline evalml/pipelines/components/transformers/column_selectors.py:128]: Mega-nit: `SelectByTypeOrTag`?  The lowercase or was bugging me

**Review by chukarsten — APPROVED** [review]
Nice!  I'd personally be a fan of upper-casing the 'or' in the class name.  I'd also like to resolve whether we should error or warn if a selection is attempted on data without the expected type present.  For some reason, our components seem a little more intentional, rather than explorational, to me, so I would expect if this component is added to select a certain type, that that certain type is there.  Also, I feel like I've run into issues with our partial dependence code that warns when mixed nan/str features are detected, but our downstream consumers don't see the error and the trouble manifests in strange ways.

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (angela97lin) [inline evalml/pipelines/components/transformers/column_selectors.py:137]: Just food for thought, but if we want to allow users to use this transformer via the name, this is a mighty long name that they'd have to not make a typo for... lolol

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (dsherry) [reply evalml/pipelines/components/transformers/column_selectors.py:128]: I think `SelectByType` would be slightly easier to read without losing much meaning. Thoughts?

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (dsherry) [inline evalml/pipelines/components/transformers/column_selectors.py:166]: If all the `transform` method for this class is doing is calling `super`, please delete it!

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (dsherry) [reply evalml/pipelines/components/transformers/column_selectors.py:153]: Good thinking @eccabay and great point @chukarsten .

I think we should raise an exception here.

Example: graph filters down to categoricals only, then passes those to OHE and to estimator. If there are no categoricals in the dataset, OHE will be passed empty DF (and possibly error out), estimator will be passed empty DF (and possibly error out) and if for some reason there weren't errors in training the model will not have any value because it didn't receive any input features! This feels to me like grounds for a clear error message from this component.

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (dsherry) [inline evalml/pipelines/components/transformers/column_selectors.py:146]: Why define this method as a no-op? Why not simply omit it?

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (dsherry) [inline evalml/pipelines/components/transformers/column_selectors.py:133]: Don't we need an `__init__` method defined for this class in order to accept the types/tag(s) to use for the filter?

**Review by dsherry — CHANGES_REQUESTED** [review]
@eccabay nice going, left a few comments.

Also, open question in case it hasn't already been discussed: should we define this functionality as a separate class (as is currently being done) or as simply a different option or way to use the existing `SelectColumns` transformer? I think defining a separate component makes sense, as long as there isn't too much code duplication.

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (eccabay) [reply evalml/pipelines/components/transformers/column_selectors.py:146]: The function is implemented in the parent class - I defined it here in order to overwrite the parent function.

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (angela97lin) [inline evalml/pipelines/components/transformers/column_selectors.py:149]: Nit pick on readability but `cols` vs `columns` as names is hard to differentiate; even just `selected_cols` could be easier to read?

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (angela97lin) [reply evalml/pipelines/components/transformers/column_selectors.py:153]: I like the idea of raising an exception or doing nothing at all:

Doing nothing at all: for the component independently, it can arguably be okay if the user wants to filter for a certain type and there happens to be no columns with that type. This is trickier with the component graph, though, as @dsherry pointed out, it is unclear what our other components do with an empty df. If our other components are able to handle empty dfs, I feel like it could be okay to do nothing: in the example above, we could try to have a separate branch for selecting categoricals for the OHE. If there are none, then the OHE doesn't do anything to the empty DF. That can be concatted with the non-categoricals and passed to the estimator. 

TLDR: safety net can be just to raise error but I feel like there's some validity in not doing anything at all /shrug

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (angela97lin) [reply evalml/pipelines/components/transformers/column_selectors.py:137]: Edit: I like @dsherry's suggestion of `SelectByType` :)

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (eccabay) [reply evalml/pipelines/components/transformers/column_selectors.py:133]: This implementation just piggybacks off of the same argument name as defined in the parent class, so there's no need for an explicit `__init__`. That being said, it may make sense to rename this argument to differentiate between column names and column types. Thoughts?

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (eccabay) [reply evalml/pipelines/components/transformers/column_selectors.py:153]: Does anyone know for certain that a downstream component would error out if passed an empty DataFrame? I ask since the default parameter for `SelectColumns` selects no columns and returns an empty DF (and we test to ensure this). Obviously, the situation is a little different if the user attempted to select a certain type and that failed, but I'm unclear how worried we should be about just passing empty DFs through pipelines.

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (eccabay) [reply evalml/pipelines/components/transformers/column_selectors.py:128]: Since the docstring mentions either logical type or semantic tag is possible, I'm fine renaming it to `SelectByType`!

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (eccabay) [reply evalml/pipelines/components/transformers/column_selectors.py:137]: Good point @angela97lin, I didn't think about that! Do you think just "Select By Type Transformer" would be concise but also informative enough?

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (freddyaboulton) [reply evalml/pipelines/components/transformers/column_selectors.py:166]: The downside is that we lose the docstring! But since this docstring isn't very specific to this class, that's ok I think

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (freddyaboulton) [reply evalml/pipelines/components/transformers/column_selectors.py:153]: Yea it would error out. Our components are not written to handle empty features so I think it's ok to raise an exception! My reasoning for raising a warning instead was that ww lets you select for types/tags that are not present and I figured that our component should not get in the way of that. I don't think we need to worry about passing empty features around. The pipelines created by AutoMLSearch don't do that so that would only happen if the user is doing something custom - we typically let users encounter failures in those cases. Making our components resilient to empty features could be an interesting project, but not one that we need to embark on now?

image

File: evalml/tests/component_tests/test_utils.py
Comment (ParthivNaresh) [inline evalml/tests/component_tests/test_utils.py:34]: Putting a pin in this after our conversation during office hours 😅

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (eccabay) [reply evalml/pipelines/components/transformers/column_selectors.py:153]: Post office hours discussion, moving forward by throwing an error if the resulting DataFrame is empty. I've filed an issue to track further discussion/work in #2554

File: evalml/tests/component_tests/test_utils.py
Comment (ParthivNaresh) [reply evalml/tests/component_tests/test_utils.py:34]: Issue filed #2555

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (angela97lin) [reply evalml/pipelines/components/transformers/column_selectors.py:137]: Yup, I think that's great 😁

**Comment by ParthivNaresh** [comment]
@eccabay I triggered `build_conda_pkg` on the PR just to verify that it [fails](https://github.com/alteryx/evalml/pull/2531/checks?check_run_id=3164589089#step:5:2763) without the bump in `n_components`. I went ahead and added +1 to `is_using_conda` and `is_running_py_39_or_above`.

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (dsherry) [reply evalml/pipelines/components/transformers/column_selectors.py:132]: Got it, thanks for explaining.

I think the most pythonic thing to do here would be to add a second argument called "column_types" which is a list of types corresponding 1x1 positionally to the list of columns.

Overriding the type definition of an argument in the inheritance structure adds complexity. I don't think it'll cause problems in this particular case, since the inheritance structure is currently quite simple. But I do think using a separate argument would effectively future-proof this API against any unanticipated changes we may want to make to these components later, and is generally a useful pattern to follow for that reason.

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (dsherry) [reply evalml/pipelines/components/transformers/column_selectors.py:145]: I see that `ColumnSelector._check_input_for_columns` does the following:
```python
        cols = self.parameters.get("columns") or []

        column_names = X.columns

        missing_cols = set(cols) - set(column_names)
        if missing_cols:
            raise ValueError(
                "Columns {} not found in input data".format(
                    ", ".join(f"'{col_name}'" for col_name in missing_cols)
                )
            )
```
And that the above is called during `ColumnSelector.fit`. And, since `SelectByType` doesn't define a `fit` method, `SelectByType.fit` will implicitly call `ColumnSelector.fit`, which is good 👍 

So, is there a reason why we should *not* also do that during `SelectByType.fit`?

I propose
* Make the change I suggested about defining an init method for `SelectByType` and accepting `column_types` as a separate arg. So the `columns` arg will have the same structure in both `ColumnSelector` and `SelectByType` init
* Delete `SelectByType._check_input_for_columns`

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (dsherry) [inline evalml/pipelines/components/transformers/column_selectors.py:150]: This is the same type of error we raise in `ColumnSelector._check_input_for_columns`. Do we need to define it in two places? I think the intent with the current layout of `ColumnSelector` methods was for `_check_input_for_columns` to do this sort of validation, and for `_modify_columns` to only include logic to make modifications (i.e. dropping columns for the `DropColumns` transformer)

I think we should define this error only in `_check_input_for_columns`.

If you want the error to include the column types for `SelectByType`, but omit the column types for `ColumnSelector` and other subclasses, I would suggest something like:
* Define a class attribute like `_VALIDATION_ERROR_FORMAT_STRING = "Columns {columns} not found in input data."`
* Override the value of `_VALIDATION_ERROR_FORMAT_STRING` in the `SelectByType` class definition to `"Columns {columns} of type {column_types} not found in input data"`
* Update `ColumnSelector._check_input_for_columns` to do
```python
raise ValueError(
                self._VALIDATION_ERROR_FORMAT_STRING.format(**self.parameters)
            )
```
* Delete `SelectByType._modify_columns`

Alternatively, we could simplify this, just delete `SelectColumns._modify_columns` and not have the error message say anything about types.

Make sense?

**Review by dsherry — APPROVED** [review]
@eccabay looks good! I left some specific suggestions on how to reduce the complexity of this code change even further. Not blocking merge, feel free to merge this PR and get another up to address, if you are interested in doing so.

**Comment by dsherry** [comment]
@ParthivNaresh thanks for the heads up. Thankfully, since we made the codecov patch check optional, and that is the one which is going red on this PR, @eccabay doesn't need me to do the merge! The codecov project check is still green here.

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (angela97lin) [inline evalml/pipelines/components/transformers/column_selectors.py:162]: Interesting... does `_check_input_for_columns` do anything here? Since we don't define it, I'm assuming it uses `ColumnSelector`'s, which has the line:
```
        cols = self.parameters.get("columns") or []
```
That returns an empty list for this component and therefore doesn't do any checks, right?

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (angela97lin) [inline evalml/pipelines/components/transformers/column_selectors.py:150]: This, and the following comment, has me questioning if it's worth subclassing `ColumnSelector` rather than just `Transformer`; it seems like we're not gaining much value from `ColumnSelector` 😂

File: evalml/tests/component_tests/test_column_selector_transformers.py
Comment (angela97lin) [inline evalml/tests/component_tests/test_column_selector_transformers.py:50]: This test doesn't actually test that a column type does not exist, but rather that a column does not exist! Since we're passing in the keyword `columns`, we trigger `self._check_input_for_columns(X)` to raise a ValueError, but don't we actually want to test the following:
```
transformer = SelectByType(column_types=["some fake column type"])
```
and see that the ValueError is raised? I think this is related to my previous comment about using `_check_input_for_columns` :)

**Review by angela97lin — COMMENTED** [review]
Thanks for cleaning this up and taking in all of our comments, Becca! This looks great 😁 

I do have some final comments before this is merged regarding `_check_input_for_columns`--I'm not certain it's doing exactly what we want right now in SelectByType. It checks for column names via the `columns` parameter, but SelectByType checks for column types, so I'm not sure we can just use it as is?

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (eccabay) [reply evalml/pipelines/components/transformers/column_selectors.py:162]: It's true that that line will return an empty list and therefore the missing column name check will do nothing - that's by design here. This new bit of code should check if there are column types and if they match though:
```
col_types = self.parameters.get("column_types")
if col_types:
       missing_cols = X.ww.select(col_types).empty
```
I agree, it does feel a little weird to do this here, I'm thinking it may make more sense to move this to `check_input_for_columns` within `SelectByType` itself.

File: evalml/pipelines/components/transformers/column_selectors.py
Comment (eccabay) [reply evalml/pipelines/components/transformers/column_selectors.py:150]: That's true, the way this is written there's not a lot of benefit to having `SelectByType` subclass this implementation of `ColumnSelector`. That being said, it still makes logical sense to have it be a subclass, since it does function to select columns. With that, I see this more as a reason to consider refactoring `ColumnSelector` rather than removing `SelectByType` as a subclass.

**Pull request merged**
